analyzer/evmverifier/evmverifier.go

```diff
@@ -0,0 +1,183 @@
+	EthAddress            ethCommon.Address
+}
+
+func (a *evmContractsVerifier) getUnverifiedContracts(ctx context.Context, runtime common.Runtime) ([]*unverifiedContract, error) {
+	var unverifiedContracts []*unverifiedContract
+	rows, err := a.target.Query(ctx, queries.RuntimeEVMUnverfiedContracts, runtime)
+	if err != nil {
+		return nil, fmt.Errorf("querying unverified contracts: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var contract unverifiedContract
+		if err = rows.Scan(
+			&contract.Addr,
+			&contract.AddrContextIdentifier,
+			&contract.AddrContextVersion,
+			&contract.AddrData,
+		); err != nil {
+			return nil, fmt.Errorf("scanning unverified contracts: %w", err)
+		}
+
+		// Compute the eth address from the preimage.
+		ethAddress, err := client.EVMEthAddrFromPreimage(contract.AddrContextIdentifier, contract.AddrContextVersion, contract.AddrData)
+		if err != nil {
+			return nil, fmt.Errorf("contract eth address: %w", err)
+		}
+		// TODO: assert eth address 20 bytes?
+		contract.EthAddress = ethCommon.BytesToAddress(ethAddress)
+
+		unverifiedContracts = append(unverifiedContracts, &contract)
+	}
+	return unverifiedContracts, nil
+}
+
+func (a *evmContractsVerifier) Start(ctx context.Context) {
+	if a.chain != common.ChainNameMainnet && a.chain != common.ChainNameTestnet {
+		a.logger.Warn("EVM contracts verifier only supports testnet/mainnet, stopping", "chain_name", a.chain)
+		return
+	}
+
+	for {
+		select {
+		case <-time.After(a.interval):
+		case <-ctx.Done():
+			a.logger.Warn("shutting down EVM contracts verifier", "reason", ctx.Err())
+			return
+		}
+
+		a.logger.Info("EVM contracts verifier iteration")
+
+		for _, runtime := range evmRuntimes {
+			// Load all non-verified contracts from the DB.
+			unverified, err := a.getUnverifiedContracts(ctx, runtime)
+			if err != nil {
+				a.logger.Error("failed to get unverified contracts", "err", err, "runtime", runtime)
+				continue
+			}
+			if len(unverified) == 0 {
+				a.logger.Debug("no unverified contracts", "runtime", runtime)
+				continue
+			}
+
+			// Query Sourcify for list of all verified contracts.
+			addresses, err := a.client.GetVerifiedContractAddresses(ctx, runtime)
+			if err != nil {
+				a.logger.Error("failed to get verified contract addresses", "err", err, "runtime", runtime)
+				continue
+			}
+			a.logger.Debug("got verified contract addresses", "runtime", runtime, "addresses", addresses)
+			if len(addresses) == 0 {
+				continue
+			}
+			// Create a lookup map of verified contract addresses.
+			sourcifyAddresses := make(map[ethCommon.Address]bool, len(addresses))
+			for _, address := range addresses {
+				sourcifyAddresses[address] = true
+			}
+
+			// Pick currently unverified contracts that are present in sourcify.
+			canBeVerified := []*unverifiedContract{}
+			for _, contract := range unverified {
+				if _, ok := sourcifyAddresses[contract.EthAddress]; ok {
+					canBeVerified = append(canBeVerified, contract)
```

Comment:
Go is just amazing. :face_with_spiral_eyes:
My granpappy computed set intersections by hand, and if it was good enough for him, then by God it's good enough for me!

Reply:
Out with the new, in with the old!